pkg/netutils/netutils.go

```diff
@@ -113,7 +89,350 @@ func SetDefaultGW(args *skel.CmdArgs, ifName string, gateways []net.IP, res *cni
+				}
+				if dst == "0.0.0.0/0" {
+					newRoutes = append(routes[:i], routes[i+1:]...)
+				}
+			}
+		}
+	}
+
+	// delete IPv6 default routes
+	if ipv6 {
+		for i, r := range newRoutes {
+			route, ok := r.(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("wrong route format: %v", r)
+			}
+			_, ok = route["dst"]
+			if ok {
+				dst, ok := route["dst"].(string)
+				if !ok {
+					return nil, fmt.Errorf("wrong dst format: %v", route["dst"])
+				}
+				if dst == "::0/0" {
+					newRoutes = append(newRoutes[:i], newRoutes[i+1:]...)
+				}
+			}
+		}
+	}
+	result["routes"] = newRoutes
+
+	return result, nil
+}
+
+func deleteDefaultGWResult020(result map[string]interface{}, ipv4, ipv6 bool) (map[string]interface{}, error) {
+	if ipv4 {
+		_, ok := result["ip4"]
+		if ok {
+			ip4, ok := result["ip4"].(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("wrong ip4 format: %v", result["ip4"])
+			}
+
+			_, ok = ip4["routes"]
+			if ok {
+				routes, ok := ip4["routes"].([]interface{})
+				if !ok {
+					return nil, fmt.Errorf("wrong ip4 routes format: %v", ip4["routes"])
+				}
+
+				newRoutes := routes
+				for i, r := range routes {
+					route, ok := r.(map[string]interface{})
+					if !ok {
+						return nil, fmt.Errorf("wrong ip4 route format: %v", r)
+					}
+
+					_, ok = route["dst"]
+					if ok {
+						dst, ok := route["dst"].(string)
+						if !ok {
+							return nil, fmt.Errorf("wrong dst format: %v", route["dst"])
+						}
+						if dst == "0.0.0.0/0" {
+							newRoutes = append(routes[:i], routes[i+1:]...)
+						}
+					}
+				}
+				ip4["routes"] = newRoutes
+			}
+		}
+	}
+
+	if ipv6 {
+		_, ok := result["ip6"]
+		if ok {
+			ip6, ok := result["ip6"].(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("wrong ip6 format: %v", result["ip6"])
+			}
+
+			_, ok = ip6["routes"]
+			if ok {
+				routes, ok := ip6["routes"].([]interface{})
+				if !ok {
+					return nil, fmt.Errorf("wrong ip6 routes format: %v", ip6["routes"])
+				}
+
+				newRoutes := routes
+				for i, r := range routes {
+					route, ok := r.(map[string]interface{})
+					if !ok {
+						return nil, fmt.Errorf("wrong ip6 route format: %v", r)
+					}
+
+					_, ok = route["dst"]
+					if ok {
+						dst, ok := route["dst"].(string)
+						if !ok {
+							return nil, fmt.Errorf("wrong dst format: %v", route["dst"])
+						}
+						if dst == "::0/0" {
+							newRoutes = append(routes[:i], routes[i+1:]...)
+						}
+					}
+				}
+				ip6["routes"] = newRoutes
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// AddDefaultGWCache updates libcni cache to add default gateway result
+func AddDefaultGWCache(cacheDir string, rt *libcni.RuntimeConf, netName string, ifName string, gw []net.IP) error {
+	cacheFile := filepath.Join(cacheDir, "results", fmt.Sprintf("%s-%s-%s", netName, rt.ContainerID, rt.IfName))
+
+	cache, err := ioutil.ReadFile(cacheFile)
+	if err != nil {
+		return err
+	}
```

Comment:
This is also used when on the `DeleteDefaultGWCache` function - as such, I think it should be exported as a function and called here.

Reply:
ping (just a suggestion you can ignore).

```golang
func AddDefaultGWCache(cacheDir string, rt *libcni.RuntimeConf, netName string, ifName string, gw []net.IP) error {
    cache, err := readCache(cacheDir, netName, rt.ContainerID, rt.IfName)
...
}

func readCache(dir string, networkName string, containerID string, ifaceName string) ([]byte, err) {
	return ioutil.ReadFile(
            filepath.Join(
                cacheDir, "results", fmt.Sprintf("%s-%s-%s", netName, rt.ContainerID, rt.IfName)))
```